### PR TITLE
[MO] - [system test] -> Resource allocation of ClusterOperator, Kafka, KafkaMirrorMaker components

### PR DESCRIPTION
**Pull request by see-quick**
Signed-off-by: morsak <xorsak02@stud.fit.vutbr.cz>

### Type of change

- Bugfix
- Enhancement / new feature
- Refactoring

### Description

Implements the first point from https://github.com/strimzi/strimzi-kafka-operator/issues/6668.

Moreover, this PR solves multiple things such as:
1. Test flakiness - some of the test cases were flaky, because when we use `shared_memory` across ∀ resources type (i.e., Kafka, KafkaConnect...) that was less prone to race conditions. In this PR we switched some components (f.e., Kafka, KafkaMIrroMaker/2) to have a specific RAM available and thus it is a little bit slower when we deploy such components.
2. Resource allocation - for resources, which allocate a lot of memory we restricted it, to as little as possible we could get. For instance, for MirrorMaker we specify `784Mi` memory and for Kafka `512Mi`. 

### Checklist

- [x] Make sure all tests pass (regression already passed 3 times)

### CODE REVIEW

**Comment by see-quick** [comment]
I see in V. pipeline failing on this test case, which should be fixed here -> https://github.com/strimzi/strimzi-kafka-operator/pull/6631.

**Comment by see-quick** [comment]
In `I.` pipeline, problem with namespace handling...

**Comment by im-konge** [comment]
Issue, which is seen in `I.` pipeline, will be fixed in #6656

**Comment by see-quick** [comment]
So I have no longer hitting issues with memory or CPU starvation. But we need to resolve the flakiness of our STs, which do not show previously because now this Minikube is little bit slower (because of setting memory for almost each component).

Here is the list:

`III. Pipeline`
```
[ERROR] Errors: 
[ERROR] io.strimzi.systemtest.operators.topic.ThrottlingQuotaST.testThrottlingQuotasDeleteTopic(ExtensionContext)
[ERROR]   Run 1: ThrottlingQuotaST.testThrottlingQuotasDeleteTopic:118 » Wait Timeout after 300...
[ERROR]   Run 2: ThrottlingQuotaST.testThrottlingQuotasDeleteTopic:134 » Wait Timeout after 300...
[ERROR]   Run 3: ThrottlingQuotaST.testThrottlingQuotasDeleteTopic:118 » Wait Timeout after 300...
[INFO] 
[WARNING] Flakes: 
[WARNING] io.strimzi.systemtest.operators.topic.ThrottlingQuotaST.testKafkaAdminTopicOperations(ExtensionContext)
[ERROR]   Run 1: ThrottlingQuotaST.testKafkaAdminTopicOperations:288 » Wait Timeout after 12000...
[ERROR]   Run 2: ThrottlingQuotaST.testKafkaAdminTopicOperations:288 » Wait Timeout after 12000...
[INFO]   Run 3: PASS
[INFO] 
[WARNING] io.strimzi.systemtest.operators.topic.ThrottlingQuotaST.testThrottlingQuotasCreateTopic(ExtensionContext)
[ERROR]   Run 1: ThrottlingQuotaST.testThrottlingQuotasCreateTopic:77 » Wait Timeout after 3000...
[INFO]   Run 2: PASS
```
-
`IV.Pipeline` 
```
[ERROR] Errors: 
[ERROR] io.strimzi.systemtest.operators.CustomResourceStatusIsolatedST.testKafkaConnectAndConnectorStatus(ExtensionContext)
[ERROR]   Run 1: CustomResourceStatusIsolatedST.testKafkaConnectAndConnectorStatus:290->assertKafkaConnectorStatus:617 » Wait
[ERROR]   Run 2: CustomResourceStatusIsolatedST.testKafkaConnectAndConnectorStatus:290->assertKafkaConnectorStatus:617 » Wait
[ERROR]   Run 3: CustomResourceStatusIsolatedST.testKafkaConnectAndConnectorStatus:290->assertKafkaConnectorStatus:617 » Wait
[INFO] 
[WARNING] Flakes: 
[WARNING] io.strimzi.systemtest.operators.topic.TopicST.testCreateTopicViaAdminClient(ExtensionContext)
[ERROR]   Run 1: TopicST.testCreateTopicViaAdminClient:173 
Expected: is <3>
     but: was <2>
[INFO]   Run 2: PASS
```
`VI.Pipeline`
```
[ERROR] Errors: 
[ERROR] io.strimzi.systemtest.mirrormaker.MirrorMaker2IsolatedST.testKMM2RollAfterSecretsCertsUpdateTLS(ExtensionContext)
[ERROR]   Run 1: MirrorMaker2IsolatedST.testKMM2RollAfterSecretsCertsUpdateTLS:1571 » Wait Time...
[ERROR]   Run 2: MirrorMaker2IsolatedST.testKMM2RollAfterSecretsCertsUpdateTLS:1577 » Wait Time...
[ERROR]   Run 3: MirrorMaker2IsolatedST.testKMM2RollAfterSecretsCertsUpdateTLS:1577 » Wait Time...
[INFO] 
[INFO] 
[ERROR] Tests run: 45, Failures: 0, Errors: 1, Skipped: 1
```

File: systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaMirrorMaker2Templates.java [hunk 1]
```diff
@@ -110,6 +112,10 @@ private static KafkaMirrorMaker2Builder defaultKafkaMirrorMaker2(KafkaMirrorMake
                 .withNewInlineLogging()
                     .addToLoggers("connect.root.logger.level", "DEBUG")
                 .endInlineLogging()
+                .withResources(new ResourceRequirementsBuilder()
+                    .addToLimits("memory", new Quantity("1Gi"))
+                    .addToRequests("memory", new Quantity("1Gi"))
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaMirrorMaker2Templates.java:117]: Setting requests to 1Gi is not a good idea for azure VMs

File: systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaMirrorMakerTemplates.java [hunk 2]
```diff
@@ -62,6 +64,10 @@ private static KafkaMirrorMakerBuilder defaultKafkaMirrorMaker(KafkaMirrorMaker
                 .withNewInlineLogging()
                     .addToLoggers("mirrormaker.root.logger", "DEBUG")
                 .endInlineLogging()
+                .withResources(new ResourceRequirementsBuilder()
+                    .addToLimits("memory", new Quantity("1Gi"))
+                    .addToRequests("memory", new Quantity("1Gi"))
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaMirrorMakerTemplates.java:69]: Same as above

**Comment by see-quick** [comment]
This [1] should fix the `IV.` and `VI` pipeline.

[1] -> https://github.com/strimzi/strimzi-kafka-operator/pull/6654/commits/310ad4fcd7acd6fde049fd74d65aa75c2ac82188

**Comment by see-quick** [comment]
So it seems that now it's working without any OOM problems. But we need to delete these last flaky test cases (I have already fixed dozens of test cases, which were flaky...):
`II.Pipeline`
```
[WARNING] Flakes: 
[WARNING] io.strimzi.systemtest.security.SecurityST.testCertRegeneratedAfterInternalCAisDeleted(ExtensionContext)
[ERROR]   Run 1: SecurityST.testCertRegeneratedAfterInternalCAisDeleted:819 Sent (100) and receive (0) message count is not equal
[INFO]   Run 2: PASS
[INFO] 
[INFO] 
[WARNING] Tests run: 20, Failures: 0, Errors: 0, Skipped: 0, Flakes: 1
```

III.Pipeline
```
[WARNING] Flakes: 
[WARNING] io.strimzi.systemtest.operators.topic.ThrottlingQuotaST.testThrottlingQuotasDuringAllTopicOperations(ExtensionContext)
[ERROR]   Run 1: ThrottlingQuotaST.testThrottlingQuotasDuringAllTopicOperations:141 » Wait Time...
[INFO]   Run 2: PASS
[INFO] 
[INFO] 
[WARNING] Tests run: 46, Failures: 0, Errors: 0, Skipped: 5, Flakes: 1
```
IV.Pipeline
```
[WARNING] Flakes: 
[WARNING] io.strimzi.systemtest.operators.CustomResourceStatusIsolatedST.testKafkaConnectAndConnectorStatus(ExtensionContext)
[ERROR]   Run 1: CustomResourceStatusIsolatedST.testKafkaConnectAndConnectorStatus:314 » Wait T...
[INFO]   Run 2: PASS
[INFO] 
[WARNING] io.strimzi.systemtest.operators.MultipleClusterOperatorsIsolatedST.testKafkaCCAndRebalanceWithMultipleCOs(ExtensionContext)
[ERROR]   Run 1: MultipleClusterOperatorsIsolatedST.testKafkaCCAndRebalanceWithMultipleCOs:217 » Wait
[INFO]   Run 2: PASS
[INFO] 
[WARNING] io.strimzi.systemtest.operators.user.UserST.testScramUserWithQuotas(ExtensionContext)
[ERROR]   Run 1: UserST.testScramUserWithQuotas:228->testUserWithQuotas:276 » KubeCluster `kube...
[INFO]   Run 2: PASS
[INFO] 
[INFO] 
[WARNING] Tests run: 60, Failures: 0, Errors: 0, Skipped: 5, Flakes: 3
```

File: systemtest/src/main/java/io/strimzi/systemtest/Constants.java [hunk 3]
```diff
@@ -394,8 +394,8 @@ public interface Constants {
      * Resource constants for Cluster Operator. In case we execute more than 5 test cases in parallel we at least these configuration
```
Comment (michalxo) [inline systemtest/src/main/java/io/strimzi/systemtest/Constants.java:394]: `we at least these` -> with at least this configuration (?)
with minimum recommended configuration specified here?

**Comment by see-quick** [comment]
Sometimes, we face this flaky test case `RollingUpdateST#testKafkaAndZookeeperScaleUpScaleDown`, where it fails on this:
```yaml
status:
  conditions:
  - type: "NotReady"
    status: "True"
    lastTransitionTime: "2022-04-28T11:25:46.400006Z"
    reason: "InvalidStateStoreException"
    message: "The state store, topic-store, may have migrated to another instance."
  observedGeneration: 1
  topicName: "my-topic-485109851-442269632"
```
I think that this is connected to https://github.com/strimzi/strimzi-kafka-operator/issues/6671 or?

**Comment by see-quick** [comment]
So ∀ regression passed `3 times in a row` without any issues (some flake remains but that's for another PR).

**Review by im-konge — APPROVED** [review]
LGTM, assuming it works.

File: systemtest/src/main/java/io/strimzi/systemtest/Environment.java [hunk 4]
```diff
@@ -244,6 +251,10 @@ public static boolean isStrimziPodSetEnabled() {
         return STRIMZI_FEATURE_GATES.contains(Constants.USE_STRIMZI_POD_SET);
     }
 
+    public static boolean isSharedMemory() {
+        return RESOURCE_ALLOCATION_STRATEGY.contains(RESOURCE_ALLOCATION_STRATEGY_DEFAULT);
+    }
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/Environment.java:256]: Javadoc would be nice

File: systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaMirrorMaker2Templates.java [hunk 5]
```diff
@@ -111,6 +113,15 @@ private static KafkaMirrorMaker2Builder defaultKafkaMirrorMaker2(KafkaMirrorMake
                     .addToLoggers("connect.root.logger.level", "DEBUG")
                 .endInlineLogging()
             .endSpec();
+
+        if (!Environment.isSharedMemory()) {
+            kmm2b.editSpec().withResources(new ResourceRequirementsBuilder()
+                .addToLimits("memory", new Quantity("784Mi"))
+                .addToRequests("memory", new Quantity("784Mi"))
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaMirrorMaker2Templates.java:120]: maybe some comment on why we use these numbers?

File: systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaMirrorMakerTemplates.java [hunk 6]
```diff
@@ -63,6 +65,16 @@ private static KafkaMirrorMakerBuilder defaultKafkaMirrorMaker(KafkaMirrorMaker
                     .addToLoggers("mirrormaker.root.logger", "DEBUG")
                 .endInlineLogging()
             .endSpec();
+
+        if (!Environment.isSharedMemory()) {
+            kmmb.editSpec().withResources(new ResourceRequirementsBuilder()
+                .addToLimits("memory", new Quantity("784Mi"))
+                .addToRequests("memory", new Quantity("784Mi"))
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaMirrorMakerTemplates.java:72]: same as above

File: systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaTemplates.java [hunk 7]
```diff
@@ -224,6 +226,38 @@ private static KafkaBuilder defaultKafka(Kafka kafka, String name, int kafkaRepl
                     .endTopicOperator()
                 .endEntityOperator()
             .endSpec();
+
+        if (!Environment.isSharedMemory()) {
+            kb.editSpec()
+                .editKafka()
+                    .withResources(new ResourceRequirementsBuilder()
+                        .addToLimits("memory", new Quantity("512Mi"))
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/templates/crd/KafkaTemplates.java:234]: sam e as above

**Review by Frawless — APPROVED** [review]

**Comment by michalxo** [comment]
Checked it earlier, gave some comment to @see-quick  already. Else LGTM
Thanks!

**Review by scholzj — APPROVED** [review]
LGTM assuming it work. But better description of the PR of what exactly it des would for sure not hurt!

**Pull request merged**